messaging/src/main/java/org/axonframework/tracing/NestingSpanFactory.java

```diff
@@ -0,0 +1,176 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.axonframework.tracing;
+
+import org.axonframework.common.BuilderUtils;
+import org.axonframework.eventhandling.EventMessage;
+import org.axonframework.messaging.Message;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.function.Supplier;
+
+/**
+ * Implementation of {@link SpanFactory} that wraps another factory and creates nested spans when handling messages
+ * instead of starting a new trace.
+ * <p>
+ * This factory includes a time limit (by default 2 minutes, but is configurable on the builder) that handling an event
+ * will become part of the dispatching trace. After this time limit it is considered its own trace. The time limit
+ * prevents replays of events from being added to the original trace, even after a longer period of time.
+ * The time limit only applies for events and does not affect commands and queries. These will always be part of the
+ * same trace.
+ *
+ * @author Mitchell Herrijgers
+ * @since 4.6.3
+ */
+public class NestingSpanFactory implements SpanFactory {
+
+    private final SpanFactory delegateSpanFactory;
+    private final Duration timeLimit;
+    private final Clock clock;
+
+    /**
+     * Creates the {@link NestingSpanFactory} based on the {@link Builder} provided.
+     * @param builder The {@link Builder} to use during construction.
+     */
+    protected NestingSpanFactory(Builder builder) {
+        this.delegateSpanFactory = builder.delegateSpanFactory;
+        this.timeLimit = builder.timeLimit;
+        this.clock = builder.clock;
+    }
+
+    /**
+     * Creates a new {@link Builder} that can build a {@link NestingSpanFactory}.
+     * @return The {@link Builder} in charge of creating a {@link NestingSpanFactory}.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public Span createRootTrace(Supplier<String> operationNameSupplier) {
+        return delegateSpanFactory.createRootTrace(operationNameSupplier);
+    }
+
+    @Override
+    public Span createHandlerSpan(Supplier<String> operationNameSupplier, Message<?> parentMessage,
+                                  boolean isChildTrace, Message<?>... linkedParents) {
+        if (!isChildTrace && messageIsWithinTimeLimit(parentMessage)) {
+            return delegateSpanFactory.createHandlerSpan(operationNameSupplier, parentMessage, true, linkedParents);
+        }
+        return delegateSpanFactory.createHandlerSpan(operationNameSupplier, parentMessage, isChildTrace, linkedParents);
+    }
+
+    private boolean messageIsWithinTimeLimit(Message<?> parentMessage) {
```

Comment:
Nit: I would expect a separate method for validating whether we're dealing with an Event Message, honestly. Right now that's tucked away within time limit validation. Furthermore, it allows you to have this method focused for `EventMessage` implementations.

Reply:
I disagree, since it's also for deadlines (which is an implementation of EventMessage). 
Putting this into the `createHandlerSpan` method seems really odd to me. I did rename the method to make more sense from a nesting perspective :)

Reply:
This comment was definitely part of my misconception of the implementation. Thanks for clarifying the intent, @Morlack. I'll resolve this comment.